tasks/messages/actorevent/task.go

```diff
@@ -0,0 +1,142 @@
+package actorevent
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-amt-ipld/v4"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lily/model"
+	messagemodel "github.com/filecoin-project/lily/model/messages"
+	visormodel "github.com/filecoin-project/lily/model/visor"
+	"github.com/filecoin-project/lily/tasks"
+	"github.com/filecoin-project/lily/tasks/messages"
+)
+
+type Task struct {
+	node tasks.DataSource
+}
+
+func NewTask(node tasks.DataSource) *Task {
+	return &Task{
+		node: node,
+	}
+}
+
+func (t *Task) ProcessTipSets(ctx context.Context, current *types.TipSet, executed *types.TipSet) (model.Persistable, *visormodel.ProcessingReport, error) {
+	ctx, span := otel.Tracer("").Start(ctx, "ProcessTipSets")
+	if span.IsRecording() {
+		span.SetAttributes(
+			attribute.String("current", current.String()),
+			attribute.Int64("current_height", int64(current.Height())),
+			attribute.String("executed", executed.String()),
+			attribute.Int64("executed_height", int64(executed.Height())),
+			attribute.String("processor", "actor_events"),
+		)
+	}
+	defer span.End()
+
+	report := &visormodel.ProcessingReport{
+		Height:    int64(current.Height()),
+		StateRoot: current.ParentState().String(),
+	}
+
+	blkMsgRect, err := t.node.TipSetMessageReceipts(ctx, current, executed)
+	if err != nil {
+		report.ErrorsDetected = fmt.Errorf("getting tipset message receipet: %w", err)
+		return nil, report, nil
+	}
+
+	var (
+		out            = make(messagemodel.ActorEventList, 0, len(blkMsgRect))
+		errorsDetected = make([]*messages.MessageError, 0, len(blkMsgRect))
+		msgsSeen       = make(map[cid.Cid]bool, len(blkMsgRect))
+	)
+
+	for _, m := range blkMsgRect {
+		select {
+		case <-ctx.Done():
+			return nil, nil, fmt.Errorf("context done: %w", ctx.Err())
+		default:
+		}
+
+		itr, err := m.Iterator()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for itr.HasNext() {
+			msg, _, rec := itr.Next()
+			if msgsSeen[msg.Cid()] {
+				continue
+			}
+			msgsSeen[msg.Cid()] = true
+
+			if rec.EventsRoot == nil {
+				continue
+			}
+
+			evtArr, err := amt.LoadAMT(ctx, t.node.Store(), *rec.EventsRoot, amt.UseTreeBitWidth(types.EventAMTBitwidth))
+			if err != nil {
+				report.ErrorsDetected = fmt.Errorf("loading actor events amt (%s): %w", *rec.EventsRoot, err)
+				return nil, report, nil
+			}
```

Comment:
If this is capable of using Lotus APIs, you can now invoke the `ChainGetEvents` JSON-RPC operation with the events root CID. It will do the AMT handling on the server side.

Reply:
Yup this has the full lotus API, in https://github.com/filecoin-project/lily/pull/1137 we call that instead.